### PR TITLE
Have ListenerRegistration::Remove block until callbacks are complete

### PR DESCRIPTION
**Pull request by wilhuff**
This fixes a test flake observed very rarely in internal testing, where the callback starts processing, checks to see if it's muted, sees that it's not, and proceeds; meanwhile the test calls remove, and deletes state to which the listener was referring and by the time the callback happens in user code the listener callback now refers to the deleted state.

### CODE REVIEW

File: Firestore/core/src/api/query_core.cc [hunk 1]
```diff
@@ -163,7 +163,7 @@ std::unique_ptr<ListenerRegistration> Query::AddSnapshotListener(
       QuerySnapshot result(firestore_, query_, std::move(snapshot),
                            std::move(metadata));
 
-      user_listener_->OnEvent(result);
+      user_listener_->OnEvent(std::move(result));
```
Comment (dconeybe) [inline Firestore/core/src/api/query_core.cc:166]: Is this actually related to the fix? Or is this just an optimization you discovered while implementing this fix?

File: Firestore/core/src/core/event_listener.h [hunk 2]
```diff
@@ -86,7 +83,19 @@ class AsyncEventListener
...
+  // PORTING NOTE: Android uses a volatile here but that's not enough in C++.
+  //
+  // In C++, the user can call `ListenerRegistration::Remove` (which calls
+  // `Mute`) and then immediately delete the state backing the listener. Using
+  // a mutex here instead of an atomic ensures that `Mute` won't return until
+  // it's safe to delete the state backing a listener. In Java this is safe
+  // because the state backing the listener is garbage collected so it doesn't
+  // matter if the mute is concurrent with a callback.
+  //
+  // Use a recursive mutex instead of `std::mutex` to avoid deadlock in the case
```
Comment (dconeybe) [inline Firestore/core/src/core/event_listener.h:95]: Is there a test case dedicated to verifying no deadlock if Remove() is called by a callback? I didn't see one, but I may have missed it. I realize that you exercise this code path in the new unit test you added but it should have a dedicated unit test being that it would be a "guarantee" that would be easy to break unintentionally.

**Comment by dconeybe** [comment]
I mean to say also, the new test case you added is very elegant. It took a bit of brain power to follow its logic, but it is beautiful once understood!

File: Firestore/core/src/core/event_listener.h (hunk 2 above)
Comment (wilhuff) [reply Firestore/core/src/core/event_listener.h:95]: I've added a test.
